### PR TITLE
Allow streaming in constant space

### LINKED ISSUE #181: Possible to stream query results, instead of loading everything into memory?
First: Thanks for a great library!

If I read the source for `DynamoResultStream` correctly, all scan and query results are loaded completely into memory, before it can be processed. We have to deal with rather large datasets and would like to prevent OOM errors. Do you have a suggestion how it would be possible to stream the results? Ideally I would like to operate on a akka `Source[Either[DynamoReadError, T], NotUsed]`

I tried to implement a custom version of `DynamoResultStream`, but weaving this class into the existing DSL is hard / almost impossible, since all classes are private.

Any pointers would help.

### LINKED ISSUE #275: Remove `DynamoResultStream`
Scanamo has logic so that the 1MB limit for the result set size is circumvented, but this applies to all queries/scans regardless of users' intent. It may be problematic in some situations (as in #181 ) and prevents running some computations in constant space.

I propose to get rid of it, or at the very least to make this an opt-in feature, i.e. by default the behaviour of Scanamo strictly follows that of DynamoDB.

What do you think @philwills ?

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Closes #275 
Closes #181 

Adds `[scan|query]To` and `[scan|query]toPaged` that thread in an effect type with monoid properties (`Alternative <: Applicative with MonoidK`). Actually we need a stronger constraint, `MonadPlus`, as this is [required in the interpreter](https://github.com/scanamo/scanamo/pull/432), but that class is missing in cats.

- [x] add default `F ~> G` for the various interpreters
- [x] add more tests
- [x] add documentation
- ~~rewrite `scan` and `query` with `[scan|query]To`~~

### CODE REVIEW

**Comment by TBonnin** [comment]
Nitpicking: it might be good to mark such PR as draft until they are ready for review. I always wonder if I should review it or not.

**Comment by regiskuckaertz** [comment]
Yeah totally... how do I do that? 😅

**Comment by TBonnin** [comment]
I don't know if you can move a PR back to draft state. https://github.blog/2019-02-14-introducing-draft-pull-requests/
Next time though

**Comment by regiskuckaertz** [comment]
Ah yeah, that's useful! Will @ you when ready ✌️

**Comment by regiskuckaertz** [comment]
I won't rewrite `scan` as it requires to have a functor for `ScanamoOpsA`, and I'm not going there in this PR.

**Comment by regiskuckaertz** [comment]
@TBonnin ready for review ✌️

**Comment by TBonnin** [comment]
why the name `scan|queryTo`? The `To` make me think there is a target/destination/something to reach. Am I missing something?

**Comment by regiskuckaertz** [comment]
how about `scanM` and `queryM`?

File: scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala [hunk 1]
```diff
@@ -43,6 +46,27 @@ private[scanamo] trait DynamoResultStream[Req, Res] {
...
+  )(implicit M: Alternative[M]): ScanamoOpsT[M, List[Either[DynamoReadError, T]]] = {
+    def streamMore(req: Req, l: Int): ScanamoOpsT[M, List[Either[DynamoReadError, T]]] =
+      exec(withLimit(l)(req)).toFreeT[M].flatMap { res =>
+        val results = items(res).map(ScanamoFree.read[T])
+        if (results.isEmpty)
+          FreeT.liftT(M.empty)
+        else {
+          val l1 = limit(req).map(_ - results.length).fold(pageSize)(Math.min(pageSize, _))
+          lastEvaluatedKey(res)
+            .filterNot(_.isEmpty || l1 <= 0)
```
Comment (TBonnin) [inline scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala]: Is the `isEmpty` condition necessary here?

File: scanamo/src/main/scala/org/scanamo/SecondaryIndex.scala [hunk 2]
```diff
@@ -43,6 +44,25 @@ sealed abstract class SecondaryIndex[V] {
...
+  /**
+    * Performs a scan with the ability to introduce effects into the computation. This is
+    * useful for huge tables when you don't want to load the whole of it in memory, but
+    * scan it page by page, with a maximum of `pageSize` items per page..
+    *
+    * @note DynamoDB will only ever return maximum 1MB of data per scan, so `pageSize` is an
+    * upper bound.
+    */
+  def scanToPaged[M[_]: Alternative](pageSize: Int): ScanamoOpsT[M, List[Either[DynamoReadError, V]]]
+
```
Comment (TBonnin) [inline scanamo/src/main/scala/org/scanamo/SecondaryIndex.scala]: Would `paginated` better than `paged`? I am not a native english speaker so feel free to ignore my comment.

**Comment by TBonnin** [comment]
I vote `scanM`, `queryM` and `ScanPaginatedM` / `queryPaginatedM`.
Up to you though

**Comment by regiskuckaertz** [comment]
All done!

File: scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala [hunk 3]
```diff
@@ -43,6 +46,27 @@ private[scanamo] trait DynamoResultStream[Req, Res] {
...
+    req: Req,
+    pageSize: Int
+  )(implicit M: Alternative[M]): ScanamoOpsT[M, List[Either[DynamoReadError, T]]] = {
+    def streamMore(req: Req, l: Int): ScanamoOpsT[M, List[Either[DynamoReadError, T]]] =
+      exec(withLimit(l)(req)).toFreeT[M].flatMap { res =>
+        val results = items(res).map(ScanamoFree.read[T])
+        if (results.isEmpty)
+          FreeT.liftT(M.empty)
+        else {
+          val l1 = limit(req).map(_ - results.length).fold(pageSize)(Math.min(pageSize, _))
```
[the remark this answers is not in the document]
Comment (TBonnin) [reply scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala]: @regiskuckaertz have you seen this comment? ^^^

File: scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala (hunk 1 above)
Comment (TBonnin) [reply scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala]: @regiskuckaertz ^^^

File: scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (regiskuckaertz) [reply scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala]: ah no, sorry 🙈 lol. you're right!

File: scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala (hunk 1 above)
Comment (regiskuckaertz) [reply scanamo/src/main/scala/org/scanamo/DynamoResultStream.scala]: No it shouldn't, let me try.

**Pull request merged**